src/backend/distributed/commands/schema_based_sharding.c

```diff
@@ -227,18 +574,226 @@ citus_internal_unregister_tenant_schema_globally(PG_FUNCTION_ARGS)
+
+	/*
+	 * It is possible that by the time we acquire the lock on schema,
+	 * concurrent DDL has removed it. We can test this by checking the
+	 * existence of schema.
+	 */
+	if (!SearchSysCacheExists1(NAMESPACEOID, ObjectIdGetDatum(schemaId)))
+	{
+		ereport(ERROR, (errcode(ERRCODE_UNDEFINED_SCHEMA),
+						errmsg("concurrent operation removed the schema")));
+	}
+
+	/* Ensure schema checks */
+	EnsureTenantSchemaAllowed(schemaId);
+	EnsureSchemaOwner(schemaId);
+
+	/* Return if the schema is already a tenant schema */
+	char *schemaName = get_namespace_name(schemaId);
+	if (IsTenantSchema(schemaId))
+	{
+		ereport(NOTICE, (errmsg("schema %s is already a tenant schema", schemaName)));
+		PG_RETURN_VOID();
+	}
+	else
+	{
+		ereport(NOTICE, (errmsg("schema %s is converted to a tenant schema",
+								schemaName)));
+	}
+
+	/* Collect all tables by their types in the schema */
+	SchemaTables *schemaTables = TablesUnderSchema(schemaId);
+
+	/* Allow only Citus local and PG local tables in the schema */
+	List *disallowedCitusTableTypes = DisAllowedTableTypesForDistributeSchema();
+	if (SchemaHasCitusTableType(schemaTables, disallowedCitusTableTypes))
+	{
+		ereport(ERROR, (errmsg("schema already has distributed tables"),
+						errhint("Undistribute distributed tables under "
+								"the schema before distributing the schema.")));
+	}
+
+	/* We check that all table relation kinds are supported by a tenant schema */
+	List *localRelationIdList = schemaTables->localRelationIdList;
+	localRelationIdList = list_concat(localRelationIdList,
+									  schemaTables->citusLocalRelationIdList);
+	EnsureTableKindsSupportedForTenantSchema(localRelationIdList);
+
+	/* Create colocation id for single shard tables */
+	uint32 colocationId = CreateTenantSchemaColocationId();
+	ColocationParam colocationParam = {
+		.colocationParamType = COLOCATE_WITH_COLOCATION_ID,
+		.colocationId = colocationId,
+	};
+
+	/* Convert each local relation to a single shard relation */
+	Oid relationId = InvalidOid;
+	List *tablesToConvert = NIL;
+	foreach_oid(relationId, localRelationIdList)
+	{
+		/*
+		 * Skip partitions as they would be distributed/undistibuted by partitioned table.
+		 *
+		 * We should filter out partitions here before distributing the schema.
+		 * Otherwise, converted partitioned table would change oid of partitions and its
+		 * partition tables would fail with oid not exist.
+		 */
+		if (PartitionTable(relationId))
+		{
+			continue;
+		}
+
+		/*
+		 * Error early before distributing the schema if concurrent drop or some alter
+		 * statements occurred on the table. Note that we already had the AccessShareLock
+		 * lock when we call PartitionTable, so after this point we are guaranteed to
+		 * prevent concurrent drop and some modification on the table.
+		 */
+		EnsureRelationExists(relationId);
+		tablesToConvert = lappend_oid(tablesToConvert, relationId);
+	}
+	foreach_oid(relationId, tablesToConvert)
```

Comment:
nit, can we have a newline after loops

```suggestion

	foreach_oid(relationId, tablesToConvert)
```